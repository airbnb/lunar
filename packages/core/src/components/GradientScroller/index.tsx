import React from 'react';
import { StyleBlock } from 'aesthetic';
import debounce from 'lodash/debounce';
import throttle from 'lodash/throttle';
import IconChevronLeft from '@airbnb/lunar-icons/lib/interface/IconChevronLeft';
import IconChevronRight from '@airbnb/lunar-icons/lib/interface/IconChevronRight';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import DirectionalIcon from '../DirectionalIcon';

const INCREMENT = 25;

export type Props = {
  /** How much to scroll when hovering above gradient. Defaults to 15% of the container width. */
  autoScrollIncrement?: number;
  /** Children to be positioned in a vertically centered fashion. */
  children?: React.ReactNode;
  /** Whether to hide the scrollbar. Doesn't prevent scroll. */
  hideScrollbar?: boolean;
  /** Whether to show arrows on the edges. */
  showArrows?: boolean;
};

export type State = {
  showStartGradient: boolean;
  showEndGradient: boolean;
};

/** A horizontal scroller with gradients on each side. */
export class GradientScroller extends React.Component<Props & WithStylesProps, State> {
  static defaultProps = {
    children: null,
    hideScrollbar: false,
    showArrows: false,
  };

  contentsRef: HTMLDivElement | null = null;

  observer: ResizeObserver;

  scrollInterval: number = 0;

  scrollerRef: HTMLDivElement | null = null;

  state = {
    showStartGradient: false,
    showEndGradient: true,
  };

  constructor(props: Props & WithStylesProps) {
    super(props);

    // Register resize observer before mounting
    this.observer = new window.ResizeObserver(this.handleObserver);
  }

  componentWillUnmount() {
    this.observer.disconnect();
  }

  calculate() {
    const scroller = this.scrollerRef;
    const contents = this.contentsRef;

    if (!scroller || !contents) {
      return;
    }

    const scrollerWidth = scroller.clientWidth;
    const maxChildWidth = this.calculateMaxChildWidth(contents);

    this.setState({
      showStartGradient: scroller.scrollLeft > 0,
      showEndGradient: maxChildWidth > scrollerWidth,
    });
  }

  calculateAutoScrollIncrement(): number {
    let increment = this.props.autoScrollIncrement;

    // 15% of the width
    if (!increment && this.scrollerRef) {
      increment = this.scrollerRef.offsetWidth * 0.15;
    }

    if (!increment) {
      increment = INCREMENT;
    }

    return increment;
  }

  calculateMaxChildWidth(contents: Element) {
    return Array.from(contents.children).reduce(
      (width, child) => Math.max(child.clientWidth, width),
      contents.clientWidth,
    );
  }

  doScroll(amount: number, interval: boolean = true) {
    const callback = () => {
      if (this.scrollerRef) {
        this.scrollerRef.scrollLeft += amount;
        this.handleScrollThrottled();
      }
    };

    this.handleScrollStop();

    if (interval) {
      this.scrollInterval = window.setInterval(callback, 100);
    } else {
      callback();
    }
  }

  private handleContentsRef = (ref: HTMLDivElement) => {
    if (ref) {
      this.observer.observe(ref);
    } else if (this.contentsRef) {
      this.observer.unobserve(this.contentsRef);
    }

    this.contentsRef = ref;
  };

  private handleScrollerRef = (ref: HTMLDivElement) => {
    if (ref) {
      this.observer.observe(ref);
    } else if (this.scrollerRef) {
      this.observer.unobserve(this.scrollerRef);
    }

    this.scrollerRef = ref;
  };

  private handleObserver = (entries: ResizeObserverEntry[]) => {
    let calculate = false;

    entries.forEach(({ contentRect, target }) => {
      const width = target.dataset.prevWidth;
      const nextWidth = String(contentRect.width);

      // Dataset stores as strings
      if (!width || nextWidth !== width) {
        calculate = true;

        // eslint-disable-next-line no-param-reassign
        target.dataset.prevWidth = nextWidth;
      }
    });

    if (calculate) {
      this.handleResizeDebounced();
    }
  };

  private handleResize = () => {
    this.calculate();
  };

  private handleResizeDebounced = debounce(this.handleResize, 150);

  private handleScroll = () => {
    const target = this.scrollerRef;
    const { showStartGradient, showEndGradient } = this.state;

    if (!target) {
      return;
    }

    if (target.scrollLeft > 0 && !showStartGradient) {
      this.setState({
        showStartGradient: true,
      });
    } else if (target.scrollLeft === 0 && showStartGradient) {
      this.setState({
        showStartGradient: false,
      });
    }

    // Get the largest width child to calculate against
    const scrolledWidth = target.scrollLeft + target.clientWidth;
    const maxChildWidth = this.calculateMaxChildWidth(target.children[0]);

    if (scrolledWidth < maxChildWidth && !showEndGradient) {
      this.setState({
        showEndGradient: true,
      });
    } else if (scrolledWidth >= maxChildWidth && showEndGradient) {
      this.setState({
        showEndGradient: false,
      });
    }
  };

  private handleScrollThrottled = throttle(this.handleScroll, 100);

  private handleScrollLeft = (event: React.MouseEvent) => {
    this.doScroll(-this.calculateAutoScrollIncrement(), event.type !== 'click');
  };

  private handleScrollRight = (event: React.MouseEvent) => {
    this.doScroll(this.calculateAutoScrollIncrement(), event.type !== 'click');
  };

  private handleScrollStop = () => {
    if (this.scrollInterval) {
      window.clearInterval(this.scrollInterval);
    }
  };

  render() {
    const { cx, children, styles, hideScrollbar, showArrows } = this.props;
    const { showStartGradient, showEndGradient } = this.state;

    return (
      <div className={cx(styles.container)}>
        <div className={cx(styles.leftGradient, showStartGradient && styles.gradient_reveal)}>
          {showArrows ? (
            <button
              className={cx(styles.leftArrow, hideScrollbar && styles.arrow_hideScrollbar)}
              type="button"
              onClick={this.handleScrollLeft}
            >
              <DirectionalIcon
                direction="left"
                left={IconChevronLeft}
                right={IconChevronRight}
                size="2em"
                decorative
              />
            </button>
          ) : (
            <span
              className={cx(styles.scrollTrigger)}
              onMouseEnter={this.handleScrollLeft}
              onMouseLeave={this.handleScrollStop}
            />
          )}
        </div>

        <div className={cx(styles.rightGradient, showEndGradient && styles.gradient_reveal)}>
          {showArrows ? (
            <button
              className={cx(styles.rightArrow, hideScrollbar && styles.arrow_hideScrollbar)}
              type="button"
              onClick={this.handleScrollRight}
            >
              <DirectionalIcon
                direction="right"
                left={IconChevronLeft}
                right={IconChevronRight}
                size="2em"
                decorative
              />
            </button>
          ) : (
            <span
              className={cx(styles.scrollTrigger)}
              onMouseEnter={this.handleScrollRight}
              onMouseLeave={this.handleScrollStop}
            />
          )}
        </div>

        <div
          className={cx(styles.scroller, hideScrollbar && styles.scroller_hideScrollbar)}
          ref={this.handleScrollerRef}
          onScroll={this.handleScrollThrottled}
        >
          <div className={cx(styles.contents)} ref={this.handleContentsRef}>
            {children}
          </div>
        </div>
      </div>
    );
  }
}

export default withStyles(({ color, unit, ui, pattern, transition }) => {
  const scrollbarHeight = unit * 1.5;

  const gradient: StyleBlock = {
    ...transition.fade,
    position: 'absolute',
    top: 0,
    bottom: 0,
    zIndex: 1,
    content: "''",
    opacity: 0,
    pointerEvents: 'none',
  };

  const arrow: StyleBlock = {
    ...pattern.resetButton,
    position: 'absolute',
    top: '45%',
    marginTop: -scrollbarHeight / 2,
    transform: 'translateY(-50%)',
    backgroundColor: color.accent.bg,
    borderRadius: '50%',
    boxShadow: ui.boxShadow,
    padding: unit,
    outline: 'none',

    ':hover': {
      backgroundColor: color.accent.bgHover,
    },
  };

  return {
    container: {
      width: '100%',
      position: 'relative',
      margin: -unit / 2,
    },

    leftGradient: {
      ...gradient,
      width: unit * 5,
      left: 0,
      background: `linear-gradient(to right, ${color.accent.bg} 40%, transparent 100%)`,
    },

    leftArrow: {
      ...arrow,
      left: -unit,
    },

    rightGradient: {
      ...gradient,
      width: unit * 5,
      right: 0,
      background: `linear-gradient(to left, ${color.accent.bg} 40%, transparent 100%)`,
    },

    rightArrow: {
      ...arrow,
      right: -unit,
    },

    arrow_hideScrollbar: {
      top: '45%',
      marginTop: 0,
    },

    gradient_reveal: {
      opacity: 1,
      pointerEvents: 'auto',
    },

    scroller: {
      overflowX: 'auto',
      overflowY: 'hidden',
      display: 'flex',
      alignItems: 'center',
      alignContent: 'stretch',
      paddingBottom: scrollbarHeight,
      '-webkit-overflow-scrolling': 'touch',
    },

    scroller_hideScrollbar: {
      paddingBottom: 0,

      '::-webkit-scrollbar': {
        display: 'none',
      },
    },

    scrollTrigger: {
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
    },

    contents: {
      flexBasis: '100%',
      padding: unit / 2,
    },
  };
})(GradientScroller);
