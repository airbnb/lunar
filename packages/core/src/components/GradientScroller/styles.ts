import { StyleBlock } from 'aesthetic';
import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color, unit, ui, pattern, transition }) => {
  const scrollbarHeight = unit * 1.5;

  const gradient: StyleBlock = {
    ...transition.fade,
    position: 'absolute',
    top: 0,
    bottom: 0,
    zIndex: 1,
    content: "''",
    opacity: 0,
    pointerEvents: 'none',
  };

  const arrow: StyleBlock = {
    ...pattern.resetButton,
    position: 'absolute',
    top: '45%',
    marginTop: -scrollbarHeight / 2,
    transform: 'translateY(-50%)',
    backgroundColor: color.accent.bg,
    borderRadius: '50%',
    boxShadow: ui.boxShadow,
    padding: unit,
    outline: 'none',

    ':hover': {
      backgroundColor: color.accent.bgHover,
    },
  };

  return {
    container: {
      width: '100%',
      position: 'relative',
      margin: -unit / 2,
    },

    leftGradient: {
      ...gradient,
      width: unit * 5,
      left: 0,
      background: `linear-gradient(to right, ${color.accent.bg} 40%, transparent 100%)`,
    },

    leftArrow: {
      ...arrow,
      left: -unit,
    },

    rightGradient: {
      ...gradient,
      width: unit * 5,
      right: 0,
      background: `linear-gradient(to left, ${color.accent.bg} 40%, transparent 100%)`,
    },

    rightArrow: {
      ...arrow,
      right: -unit,
    },

    arrow_hideScrollbar: {
      top: '45%',
      marginTop: 0,
    },

    gradient_reveal: {
      opacity: 1,
      pointerEvents: 'auto',
    },

    scroller: {
      overflowX: 'auto',
      overflowY: 'hidden',
      display: 'flex',
      alignItems: 'center',
      alignContent: 'stretch',
      paddingBottom: scrollbarHeight,
      '-webkit-overflow-scrolling': 'touch',
    },

    scroller_hideScrollbar: {
      paddingBottom: 0,

      '::-webkit-scrollbar': {
        display: 'none',
      },
    },

    scrollTrigger: {
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
    },

    contents: {
      flexBasis: '100%',
      padding: unit / 2,
    },
  };
};

export default styleSheet;

export { styleSheet };
