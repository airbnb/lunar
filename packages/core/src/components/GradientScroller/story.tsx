import React from 'react';
import LoremIpsum from ':storybook/components/LoremIpsum';
import Text from '../Text';
import Button from '../Button';
import ButtonGroup from '../ButtonGroup';
import GradientScroller from '.';

class ButtonGroupDemo extends React.Component<{}, { buttons: number[] }> {
  state = { buttons: [] };

  handleClick = () => {
    this.setState(prevState => ({
      buttons: [...prevState.buttons, prevState.buttons.length],
    }));
  };

  render() {
    return (
      <div style={{ borderBottom: '2px solid #DBDBDB', paddingBottom: 4 }}>
        <ButtonGroup>
          <Button onClick={this.handleClick}>Add</Button>

          {this.state.buttons.map(count => (
            <Button key={count}>{count}</Button>
          ))}
        </ButtonGroup>
      </div>
    );
  }
}

export default {
  title: 'Core/GradientScroller',
  parameters: {
    inspectComponents: [GradientScroller],
  },
};

export function aScrollableContainer() {
  return (
    <GradientScroller>
      <Text>
        <span style={{ whiteSpace: 'nowrap' }}>
          <LoremIpsum />
          <LoremIpsum />
        </span>
      </Text>
    </GradientScroller>
  );
}

aScrollableContainer.story = {
  name: 'A scrollable container.',
};

export function withLeftAndRightArrows() {
  return (
    <GradientScroller showArrows hideScrollbar>
      <Text>
        <span style={{ whiteSpace: 'nowrap' }}>
          <LoremIpsum />
          <LoremIpsum />
        </span>
      </Text>
    </GradientScroller>
  );
}

withLeftAndRightArrows.story = {
  name: 'With left and right arrows.',
};

export function withNoScrollbarAndVariableWidthChildren() {
  return (
    <GradientScroller hideScrollbar>
      <ButtonGroupDemo />
    </GradientScroller>
  );
}

withNoScrollbarAndVariableWidthChildren.story = {
  name: 'With no scrollbar and variable width children.',
  parameters: { happo: false },
};
