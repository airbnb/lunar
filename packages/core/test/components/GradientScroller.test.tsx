import React from 'react';
import { shallowWithStyles, mockResizeObserver } from '@airbnb/lunar-test-utils';
import GradientScroller, {
  GradientScroller as BaseGradientScroller,
} from '../../src/components/GradientScroller';

describe('<GradientScroller />', () => {
  let unmockObserver: () => void;

  beforeEach(() => {
    unmockObserver = mockResizeObserver();
  });

  afterEach(() => {
    unmockObserver();
  });

  it('has the correct default sizing', () => {
    const wrapper = shallowWithStyles(<GradientScroller />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders arrows when `showArrows` is true', () => {
    const wrapper = shallowWithStyles(<GradientScroller showArrows />);

    expect(wrapper.find('button')).toHaveLength(2);

    wrapper.setProps({
      showArrows: false,
    });

    expect(wrapper.find('button')).toHaveLength(0);
  });

  it('triggers scroll when left arrow is clicked', () => {
    const spy = jest.fn();
    const wrapper = shallowWithStyles(<GradientScroller showArrows />);

    (wrapper.instance() as BaseGradientScroller).doScroll = spy;

    wrapper
      .find('button')
      .at(0)
      .simulate('click', { type: 'click' });

    expect(spy).toHaveBeenCalled();
  });

  it('triggers scroll when right arrow is clicked', () => {
    const spy = jest.fn();
    const wrapper = shallowWithStyles(<GradientScroller showArrows />);

    (wrapper.instance() as BaseGradientScroller).doScroll = spy;

    wrapper
      .find('button')
      .at(1)
      .simulate('click', { type: 'click' });

    expect(spy).toHaveBeenCalled();
  });

  describe('calculateAutoScrollIncrement()', () => {
    it('uses prop value', () => {
      const wrapper = shallowWithStyles(<GradientScroller autoScrollIncrement={10} />);

      expect((wrapper.instance() as BaseGradientScroller).calculateAutoScrollIncrement()).toBe(10);
    });

    it('uses ref width', () => {
      const wrapper = shallowWithStyles(<GradientScroller />);
      const instance = wrapper.instance() as BaseGradientScroller;
      const ref = document.createElement('div');

      Object.defineProperty(ref, 'offsetWidth', {
        value: 100,
      });

      instance.scrollerRef = ref;

      expect(instance.calculateAutoScrollIncrement()).toBe(15);
    });

    it('uses fallback constant', () => {
      const wrapper = shallowWithStyles(<GradientScroller />);
      const instance = wrapper.instance() as BaseGradientScroller;

      instance.scrollerRef = null;

      expect(instance.calculateAutoScrollIncrement()).toBe(25);
    });
  });
});
