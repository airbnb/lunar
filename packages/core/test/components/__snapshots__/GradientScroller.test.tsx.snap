// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GradientScroller /> has the correct default sizing 1`] = `
<div
  className="container_96a0ch"
>
  <div
    className="leftGradient_n8tqy6"
  >
    <span
      className="scrollTrigger_1mfsr54"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    />
  </div>
  <div
    className="rightGradient_18cxrym-o_O-gradient_reveal_18smsgc"
  >
    <span
      className="scrollTrigger_1mfsr54"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    />
  </div>
  <div
    className="scroller_cb33bt"
    onScroll={[Function]}
  >
    <div
      className="contents_lrf9wj"
    />
  </div>
</div>
`;
